src/tslint.ts

```diff
@@ -49,95 +45,25 @@ class Linter {
+        return MultiLinter.getFileNames(program);
     }
 
     constructor(private fileName: string, private source: string, options: ILinterOptionsRaw, private program?: ts.Program) {
-        this.options = this.computeFullOptions(options);
+       this.options = this.computeFullOptions(options);
     }
 
     public lint(): LintResult {
-        const failures: RuleFailure[] = [];
-        let sourceFile: ts.SourceFile;
-        if (this.program) {
-            sourceFile = this.program.getSourceFile(this.fileName);
-            // check if the program has been type checked
-            if (!("resolvedModules" in sourceFile)) {
-                throw new Error("Program must be type checked before linting");
-            }
-        } else {
-            sourceFile = getSourceFile(this.fileName, this.source);
-        }
-
-        // walk the code first to find all the intervals where rules are disabled
-        const rulesWalker = new EnableDisableRulesWalker(sourceFile, {
-            disabledIntervals: [],
-            ruleName: "",
-        });
-        rulesWalker.walk(sourceFile);
-        const enableDisableRuleMap = rulesWalker.enableDisableRuleMap;
-
-        const rulesDirectories = this.options.rulesDirectory;
-        const configuration = this.options.configuration.rules;
-        const configuredRules = loadRules(configuration, enableDisableRuleMap, rulesDirectories);
-        const enabledRules = configuredRules.filter((r) => r.isEnabled());
-        for (let rule of enabledRules) {
-            let ruleFailures: RuleFailure[] = [];
-            if (this.program && rule instanceof TypedRule) {
-                ruleFailures = rule.applyWithProgram(sourceFile, this.program);
-            } else {
-                ruleFailures = rule.apply(sourceFile);
-            }
-            for (let ruleFailure of ruleFailures) {
-                if (!this.containsRule(failures, ruleFailure)) {
-                    failures.push(ruleFailure);
-                }
-            }
-        }
-
-        let formatter: IFormatter;
-        const formattersDirectory = getRelativePath(this.options.formattersDirectory);
-
-        const Formatter = findFormatter(this.options.formatter, formattersDirectory);
-        if (Formatter) {
-            formatter = new Formatter();
-        } else {
-            throw new Error(`formatter '${this.options.formatter}' not found`);
-        }
-
-        const output = formatter.format(failures);
-        return {
-            failureCount: failures.length,
-            failures: failures,
-            format: this.options.formatter,
-            output: output,
-        };
-    }
-
-    private containsRule(rules: RuleFailure[], rule: RuleFailure) {
-        return rules.some((r) => r.equals(rule));
+        const multiLinter: MultiLinter = new MultiLinter(this.options, this.program);
+        multiLinter.lint(this.fileName, this.source, this.options.configuration);
+        return multiLinter.getResult();
```

Comment:
The idea is to separate instance options from (optional) per file configurations.